main.c

```diff
@@ -137,7 +142,58 @@ static void linux_boot(const struct boot_params boot_params)
     printf("  pref_address: 0x%llx\n", l_params->pref_address);
     printf("\n");
 
-    char *linux_header_end = (char *)(boot_params.kernel_addr + 0x201);
+    // We decided that the loader moves all kernels to their preferred address, even if they
+    // are relocatable.
+    uintptr_t actual_kernel_load_addr = boot_params.kernel_addr;
+
+    // Checks if a relocation of the kernel is required
+    if (boot_params.kernel_addr == actual_kernel_load_addr)
+    {
+        printf("The kernel is already loaded at its preferred address.\n");
+        printf("Booting kernel at 0x%lx\n", actual_kernel_load_addr);
+    } else {
+        printf("The kernel will be moved to its preferred load address.\n");
+        printf(
+            "Relocating from 0x%lx to %x%lx\n",
+            boot_params.kernel_addr,
+            actual_kernel_load_addr
+        );
+
+        uintptr_t kernel_begin_addr = actual_kernel_load_addr;
+        uintptr_t kernel_end_addr = kernel_begin_addr + l_params->init_size;
```

Comment:
in general more const would be nice because it reduces the amount of cognitive considerations while reading.